signature/src/lib.rs

```diff
@@ -30,3 +35,377 @@ pub extern "C" fn verify_rsa(
+HDAaBgNVBAoTE01vemlsbGEgQ29ycG9yYXRpb24xLzAtBgNVBAsTJk1vemlsbGEg
+QU1PIFByb2R1Y3Rpb24gU2lnbmluZyBTZXJ2aWNlMR8wHQYDVQQDExZyb290LWNh
+LXByb2R1Y3Rpb24tYW1vggEBMB0GA1UdDgQWBBSzvOpYdKvhbngqsqucIx6oYyyX
+tzANBgkqhkiG9w0BAQwFAAOCAgEAaNSRYAaECAePQFyfk12kl8UPLh8hBNidP2H6
+KT6O0vCVBjxmMrwr8Aqz6NL+TgdPmGRPDDLPDpDJTdWzdj7khAjxqWYhutACTew5
+eWEaAzyErbKQl+duKvtThhV2p6F6YHJ2vutu4KIciOMKB8dslIqIQr90IX2Usljq
+8Ttdyf+GhUmazqLtoB0GOuESEqT4unX6X7vSGu1oLV20t7t5eCnMMYD67ZBn0YIU
+/cm/+pan66hHrja+NeDGF8wabJxdqKItCS3p3GN1zUGuJKrLykxqbOp/21byAGog
+Z1amhz6NHUcfE6jki7sM7LHjPostU5ZWs3PEfVVgha9fZUhOrIDsyXEpCWVa3481
+LlAq3GiUMKZ5DVRh9/Nvm4NwrTfB3QkQQJCwfXvO9pwnPKtISYkZUqhEqvXk5nBg
+QCkDSLDjXTx39naBBGIVIqBtKKuVTla9enngdq692xX/CgO6QJVrwpqdGjebj5P8
+5fNZPABzTezG3Uls5Vp+4iIWVAEDkK23cUj3c/HhE+Oo7kxfUeu5Y1ZV3qr61+6t
+ZARKjbu1TuYQHf0fs+GwID8zeLc2zJL7UzcHFwwQ6Nda9OJN4uPAuC/BKaIpxCLL
+26b24/tRam4SJjqpiq20lynhUrmTtt6hbG3E1Hpy3bmkt2DYnuMFwEx2gfXNcnbT
+wNuvFqc=
+-----END CERTIFICATE-----";
+    const PROD_ROOT_HASH: &str = "97e8ba9cf12fb3de53cc42a4e6577ed64df493c247b414fea036818d3823560e";
+    const PROD_HOSTNAME: &str = "aus.content-signature.mozilla.org";
+    // Fetched from: https://aus5.mozilla.org/json/1/FirefoxVPN/2.14.0/WINNT_x86_64/release-cdntest/update.json
+    const PROD_SIGNATURE: &str = "znYFqdKKFgijVgUhnq5VuZxtI5Zay8MARVFr3cG1CbB9eH9slQFkE9ZjMdLzbf5OZqj2gds1OqbCm45L38e2joKD_mCAUGtajebztDdWx9Rqgmn-9vu6t-SCl6HQrzbh";
+    const PROD_INPUT_DATA: &[u8] =
+        b"{\"version\": \"2.17.0\", \"url\": \"https://archive.mozilla.org/pub/vpn/candidates/2.17.0-candidates/build20230926170708/windows/MozillaVPN.msi\", \"required\": false, \"hashFunction\": \"sha512\", \"hashValue\": \"97fe0255e50cd33d3ed50fe94edf3d95b72950b28f9b0157b43195a6c190aaf7e34543044c14d42deabf97b84910da1fbb7f752eaf800130315c2ad5172ea45c\"}";
+
+    #[test]
+    fn test_verify_succeeds_if_valid() {
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            1615559719, // March 12, 2021
+            ROOT_HASH,
+            VALID_HOSTNAME,
+        );
+        assert!(r.is_ok(), "Found unexpected error: {}", r.unwrap_err());
+    }
+
+    #[test]
+    fn test_verify_fails_if_root_hash_mismatch() {
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            1615559719, // March 12, 2021
+            &ROOT_HASH.replace("A", "B"),
+            VALID_HOSTNAME,
+        );
+        assert_eq!(r, Err(BalrogError::RootHashMismatch));
+    }
+
+    #[test]
+    fn test_verify_fails_if_cert_invalid() {
+        let r = parse_and_verify(
+            INVALID_CERTIFICATE,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            1615559719, // March 12, 2021
+            ROOT_HASH,
+            VALID_HOSTNAME,
+        );
+        /* I think this is the wrong error code actually...  */
+        /* I would have expected X509Error::InvalidCertificate */
```

Comment:
Does this need more investigation? Or can this be checked in with this behavior?

Reply:
I dug into it, and it turns out that the example PEM file from that I stole from `rc_crypto` for this test contained invalid base64 padding. Fixing the padding instead threw a DER encoding error, which I suppose is a sensible error to get in this case.